ui/ducks/send/send.js

```diff
@@ -0,0 +1,1325 @@
+      slice.caseReducers.validateGasField(state);
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    /**
+     * sets the provided gasPrice in state and then recomputes the gasTotal,
+     * setting that value in state as well. Once gasTotal is computed
+     * recomputes the maximum amount if the current amount mode is 'MAX' and
+     * sending the native token. ERC20 assets max amount is unaffected by
+     * gasTotal so does not need to be recomputed. Finally, validates the gas
+     * field and send state, then updates the draft transaction.
+     */
+    updateGasPrice: (state, action) => {
+      state.gas.gasPrice = addHexPrefix(action.payload);
+      state.gas.gasTotal = addHexPrefix(
+        calcGasTotal(state.gas.gasLimit, state.gas.gasPrice),
+      );
+      if (
+        state.amount.mode === AMOUNT_MODES.MAX &&
+        state.asset.type === ASSET_TYPES.NATIVE
+      ) {
+        slice.caseReducers.updateAmountToMax(state);
+      }
+      slice.caseReducers.validateGasField(state);
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    /**
+     * sets the amount mode to the provided value as long as it is one of the
+     * supported modes (MAX|INPUT)
+     */
+    updateAmountMode: (state, action) => {
+      if (Object.values(AMOUNT_MODES).includes(action.payload)) {
+        state.amount.mode = action.payload;
+      }
+    },
+    updateAsset: (state, action) => {
+      state.asset.type = action.payload.type;
+      state.asset.balance = action.payload.balance;
+      if (state.asset.type === ASSET_TYPES.TOKEN) {
+        state.asset.details = action.payload.details;
+      } else {
+        // clear the details object when sending native currency
+        state.asset.details = null;
+        if (state.recipient.error === CONTRACT_ADDRESS_ERROR) {
+          // Errors related to sending tokens to their own contract address
+          // are no longer valid when sending native currency.
+          state.recipient.error = null;
+        }
+
+        if (state.recipient.warning === KNOWN_RECIPIENT_ADDRESS_ERROR) {
+          // Warning related to sending tokens to a known contract address
+          // are no longer valid when sending native currency.
+          state.recipient.warning = null;
+        }
+      }
+      // if amount mode is MAX update amount to max of new asset, otherwise set
+      // to zero. This will revalidate the send amount field.
+      if (state.amount.mode === AMOUNT_MODES.MAX) {
+        slice.caseReducers.updateAmountToMax(state);
+      } else {
+        slice.caseReducers.updateSendAmount(state, { payload: '0x0' });
+      }
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      // update the draft transaction
+      slice.caseReducers.updateDraftTransaction(state);
```

Comment:
These are called in a pair very often, maybe an opportunity for a shared function.

Reply:
I actually moved the call to updateDraftTransaction into the validateSendState method. I can't think of any time we'd want to validate the form that we don't want to regenerate the draft.